File: model/src/main/kotlin/net/dungeonhub/model/carry_queue/CarryQueueModel.java
package net.dungeonhub.model.carry_queue;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;
import me.taubsie.dungeonhub.common.DungeonHubService;
import me.taubsie.dungeonhub.common.entity.model.Model;
import me.taubsie.dungeonhub.common.enums.QueueStep;
import me.taubsie.dungeonhub.common.model.carry_difficulty.CarryDifficultyModel;
import me.taubsie.dungeonhub.common.model.carry_tier.CarryTierModel;
import me.taubsie.dungeonhub.common.model.carry_type.CarryTypeModel;
import me.taubsie.dungeonhub.common.model.discord_user.DiscordUserModel;

import java.time.Instant;

@AllArgsConstructor
@Getter
@Setter
public class CarryQueueModel implements Model {
    private final long id;
    private QueueStep queueStep;
    private DiscordUserModel carrier;
    private DiscordUserModel player;
    private Long amount;
    private CarryDifficultyModel carryDifficulty;
    private Long relationId;
    private String attachmentLink;
    private Instant time;

    public static CarryQueueModel fromJson(String json) {
        return DungeonHubService.getInstance()
                .getGson()
                .fromJson(json, CarryQueueModel.class);
    }

    public CarryTypeModel getCarryType() {
        return getCarryTier().getCarryType();
    }

    public CarryTierModel getCarryTier() {
        return getCarryDifficulty().getCarryTier();
    }

    public long calculateScore() {
        return getScoreMultiplier() * getAmount();
    }

    private long getScoreMultiplier() {
        return getCarryDifficulty().getScore();
    }
}

File: model/src/main/kotlin/net/dungeonhub/model/carry_queue/CarryQueueModel.kt
package net.dungeonhub.model.carry_queue

import net.dungeonhub.enums.QueueStep
import net.dungeonhub.model.carry_difficulty.CarryDifficultyModel
import net.dungeonhub.model.discord_user.DiscordUserModel
import net.dungeonhub.service.MoshiService
import net.dungeonhub.structure.model.UpdateableModel
import java.time.Instant

class CarryQueueModel(
    val id: Long,
    val queueStep: QueueStep,
    val carrier: DiscordUserModel,
    val player: DiscordUserModel,
    val amount: Long,
    val carryDifficulty: CarryDifficultyModel,
    val relationId: Long?,
    val attachmentLink: String?,
    val time: Instant?
) : UpdateableModel<CarryQueueUpdateModel, CarryQueueModel> {
    val carryTier = carryDifficulty.carryTier

    val carryType = carryTier.carryType

    fun calculateScore(): Long {
        return scoreMultiplier * amount
    }

    private val scoreMultiplier = carryDifficulty.score

    companion object {
        fun fromJson(json: String): CarryQueueModel {
            return MoshiService.moshi.adapter(CarryQueueModel::class.java).fromJson(json)!!
        }
    }

    override fun getUpdateModel(): CarryQueueUpdateModel {
        return CarryQueueUpdateModel(null, null, null, null, null, null, null, null, null)
    }
}